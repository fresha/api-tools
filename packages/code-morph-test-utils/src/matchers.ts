import {
  MatcherHintOptions,
  matcherHint,
  printDiffOrStringify,
  printExpected,
  printReceived,
  stringify,
} from 'jest-matcher-utils';
import * as prettier from 'prettier';

import { poorMansElixirFormat } from './elixir';

const EXPECTED_LABEL = 'Expected';
const RECEIVED_LABEL = 'Received';

declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace, @typescript-eslint/no-unused-vars
  namespace jest {
    interface Matchers<R> {
      toHaveFormattedElixirText(test: string): R;
      toHaveFormattedTypeScriptText(text: string): R;
    }
  }
}

type SourceFile = {
  getText(): string;
};

const compareText = (
  received: string,
  expected: string,
  matcherName: string,
): jest.CustomMatcherResult => {
  const matcherOptions: MatcherHintOptions = {
    comment: 'deep equality',
  };
  const pass = received === expected;

  const message = pass
    ? () =>
        // eslint-disable-next-line prefer-template
        matcherHint(matcherName, undefined, undefined, matcherOptions) +
        '\n\n' +
        `Expected: not ${printExpected(expected)}\n` +
        (stringify(expected) !== stringify(received)
          ? `Received:     ${printReceived(received)}`
          : '')
    : () =>
        // eslint-disable-next-line prefer-template
        matcherHint(matcherName, undefined, undefined, matcherOptions) +
        '\n\n' +
        printDiffOrStringify(expected, received, EXPECTED_LABEL, RECEIVED_LABEL, true);

  return { message, pass };
};

export const createMatchers = () => ({
  toHaveFormattedElixirText: (
    sourceFile: SourceFile,
    expectedText: string,
  ): jest.CustomMatcherResult => {
    return compareText(
      poorMansElixirFormat(sourceFile.getText()),
      poorMansElixirFormat(expectedText),
      'toHaveFormattedElixirText',
    );
  },

  toHaveFormattedTypeScriptText: (
    sourceFile: SourceFile,
    expectedText: string,
  ): jest.CustomMatcherResult => {
    const options: prettier.Options = {
      parser: 'typescript',
      singleQuote: true,
      trailingComma: 'all',
    };
    const received = prettier.format(sourceFile.getText(), options);
    const expected = prettier.format(expectedText, options);

    return compareText(received, expected, 'toHaveFormattedTypeScriptText');
  },
});
