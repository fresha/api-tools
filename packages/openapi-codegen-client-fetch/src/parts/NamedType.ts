import type { ActionContext } from '../context';
import type { SchemaModel } from '@fresha/openapi-model/build/3.0.3';

export abstract class NamedType {
  readonly context: ActionContext;
  readonly name: string;
  readonly schema: SchemaModel;
  readonly isRequestBody: boolean;
  readonly isRequired: boolean;

  constructor(
    context: ActionContext,
    name: string,
    schema: SchemaModel,
    isRequestBody: boolean,
    isRequired: boolean,
  ) {
    this.context = context;
    this.name = name;
    this.schema = schema;
    this.isRequestBody = isRequestBody;
    this.isRequired = isRequired;
  }

  abstract collectData(namedTypes: Map<string, NamedType>): void;
  abstract generateCode(generatedTypes: Set<string>): void;
}
