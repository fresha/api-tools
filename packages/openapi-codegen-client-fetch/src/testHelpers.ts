// eslint-disable-next-line import/no-extraneous-dependencies
import { createTSProjectTestContext } from '@fresha/openapi-codegen-test-utils';

import type { ActionContext, Context } from './context';
import type { OpenAPIModel, OperationModel } from '@fresha/openapi-model/build/3.0.3';

export const createTestContext = (openapi: OpenAPIModel): Context => {
  const base = createTSProjectTestContext(openapi);

  return {
    ...base,
    includeDeprecated: false,
    includedTags: new Set<string>(),
    excludedTags: new Set<string>(),
    transformRequest: 'kebab',
    transformResponse: 'camel',
  };
};

export const createActionTestContext = (
  operation: OperationModel,
  fileName: string,
): ActionContext => {
  const base = createTestContext(operation.root);
  const sourceFile = base.project.createSourceFile(fileName, '');

  return {
    ...base,
    sourceFile,
    operation,
  };
};
