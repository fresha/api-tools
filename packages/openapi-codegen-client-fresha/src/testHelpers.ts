// eslint-disable-next-line import/no-extraneous-dependencies
import { createTSProjectTestContext } from '@fresha/openapi-codegen-test-utils';
import { getAPIName, getRootUrlOrThrow } from '@fresha/openapi-codegen-utils';

import type { Context } from './context';
import type { OpenAPIModel } from '@fresha/openapi-model/build/3.0.3';

export const createTestContext = (openapi: OpenAPIModel, rootDir = '/'): Context => {
  const base = createTSProjectTestContext(openapi, rootDir);

  let apiRootUrl: string;
  try {
    apiRootUrl = getRootUrlOrThrow(openapi);
  } catch (err) {
    apiRootUrl = 'API_URL';
  }

  return {
    ...base,
    apiName: getAPIName(openapi),
    apiRootUrl,
  };
};
