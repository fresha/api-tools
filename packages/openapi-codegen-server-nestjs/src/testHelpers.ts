import { Nullable } from '@fresha/api-tools-core';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createTSProjectTestContext } from '@fresha/openapi-codegen-test-utils';
import { OpenAPIFactory, SchemaModel } from '@fresha/openapi-model/build/3.0.3';

import { Context } from './context';
import { Generator } from './Generator';

export const createTestContext = (nestApp = 'web', outputPath = '/tmp/'): Context => {
  const openapi = OpenAPIFactory.create();
  const base = createTSProjectTestContext(openapi, outputPath);

  base.project.createSourceFile(
    `${outputPath}/${nestApp}.module.ts`,
    `import { Module } from '@nestjs/common';
    import { AppController } from './app.controller';
    import { AppService } from './app.service';

    @Module({
      imports: [],
      controllers: [AppController],
      providers: [AppService],
    })
    export class AppModule {}`,
  );

  return {
    ...base,
    nestApp,
    addDTO: jest.fn(),
  };
};

export const createGenerator = (nestApp = 'app', rootDir = '/'): Generator => {
  const context = createTestContext(nestApp, rootDir);

  let generator: Generator;
  // eslint-disable-next-line prefer-const
  generator = new Generator({
    ...context,
    addDTO: (name: string, schema: Nullable<SchemaModel>) => generator.addDTO(name, schema),
  });

  return generator;
};
