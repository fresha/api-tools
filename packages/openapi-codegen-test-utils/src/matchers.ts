import {
  MatcherHintOptions,
  matcherHint,
  printDiffOrStringify,
  printExpected,
  printReceived,
  stringify,
} from 'jest-matcher-utils';
import * as prettier from 'prettier';

import type { SourceFile } from 'ts-morph';

const EXPECTED_LABEL = 'Expected';
const RECEIVED_LABEL = 'Received';

// eslint-disable-next-line @typescript-eslint/no-namespace, @typescript-eslint/no-unused-vars
declare global {
  namespace jest {
    interface Matchers<R> {
      toHaveFormattedText(text: string): R;
    }
  }
}

export const createMatchers = () => {
  expect.extend({
    toHaveFormattedText: (sourceFile: SourceFile, expectedText: string): jest.CustomMatcherResult => {
      const options: prettier.Options = {
        parser: 'typescript',
        singleQuote: true,
        trailingComma: 'all',
      };
      const received = prettier.format(sourceFile.getText(), options);
      const expected = prettier.format(expectedText, options);

      const matcherName = 'toHaveFormattedText';
      const matcherOptions: MatcherHintOptions = {
        comment: 'deep equality',
      };
      const pass = received === expected;

      const message = pass
        ? () =>
            // eslint-disable-next-line prefer-template
            matcherHint(matcherName, undefined, undefined, matcherOptions) +
            '\n\n' +
            `Expected: not ${printExpected(expected)}\n` +
            (stringify(expected) !== stringify(received)
              ? `Received:     ${printReceived(received)}`
              : '')
        : () =>
            // eslint-disable-next-line prefer-template
            matcherHint(matcherName, undefined, undefined, matcherOptions) +
            '\n\n' +
            printDiffOrStringify(expected, received, EXPECTED_LABEL, RECEIVED_LABEL, true);

      return { message, pass };
    },
  });
};
