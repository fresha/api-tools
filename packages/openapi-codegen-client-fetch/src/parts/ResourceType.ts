import { addImportDeclaration, addTypeAlias } from '@fresha/code-morph-ts';
import { assert, getSchemaMultiProperty } from '@fresha/openapi-codegen-utils';
import { StructureKind, SyntaxKind } from 'ts-morph';

import { NamedType } from './NamedType';
import { schemaToType } from './utils';

import type { ActionContext } from '../context';
import type { JSONValue, Nullable } from '@fresha/api-tools-core';
import type { SchemaModel } from '@fresha/openapi-model/build/3.0.3';

enum RelationshipCardinality {
  ZeroOrOne = '0',
  One = '1',
  Many = 'N',
}

type RelationshipInfo = {
  resourceType: string;
  cardinality: RelationshipCardinality;
  required: boolean;
};

export class ResourceType extends NamedType {
  protected isRequestBody: boolean;
  protected resourceType: Nullable<string>;
  protected attributesSchema: Nullable<SchemaModel>;
  protected relationships: Map<string, RelationshipInfo>;

  constructor(context: ActionContext, name: string, schema: SchemaModel, isRequestBody: boolean) {
    super(context, name, schema);
    this.isRequestBody = isRequestBody;
    this.resourceType = null;
    this.attributesSchema = null;
    this.relationships = new Map<string, RelationshipInfo>();
  }

  collectData(_namedTypes: Map<string, NamedType>): void {
    if (this.schema == null) {
      return;
    }

    this.collectAttributesData();

    const relationshipsSchema = this.schema.getPropertyDeep('relationships');
    assert(
      !relationshipsSchema || relationshipsSchema.type === 'object',
      `Resource relationships schema is present, but it is not an object (it is ${String(
        relationshipsSchema?.type,
      )})`,
      this.context.operation,
    );
    if (relationshipsSchema) {
      this.collectRelationshipsData(relationshipsSchema);
    }
  }

  private collectAttributesData(): void {
    const typeSchema = this.schema.getPropertyDeep('type');

    if (!typeSchema) {
      this.context.logger.warn(
        `Cannot find type field in schema ${this.context.operation.parent.pathUrl}`,
      );
      // this.context.console.log(this.schema);
      return;
    }

    assert(typeSchema, `Missing 'type' property in resource ${this.name}`, this.context.operation);

    if (typeSchema.enum?.length !== 1) {
      this.context.logger.warn(
        `Expected resource schema to have only one allowed value, got ${String(
          typeSchema.enum?.length,
        )} ${this.context.operation.parent.pathUrl}`,
      );
      // this.context.console.log(this.schema);
      return;
    }

    assert(
      typeSchema.enum?.length === 1,
      `Expected resource schema to have only one allowed value, got ${String(
        typeSchema.enum?.length,
      )}`,
      this.context.operation,
    );

    const resourceType = typeSchema.enum[0];
    assert(
      typeof resourceType === 'string',
      `Expected resource type to be a string, got ${typeof resourceType}`,
      this.context.operation,
    );

    this.resourceType = resourceType;

    const attributesSchema = this.schema.getPropertyDeep('attributes') ?? null;
    assert(
      !attributesSchema || attributesSchema.type === 'object',
      `Resource attributes schema is present, but it is not an object (it is ${String(
        attributesSchema?.type,
      )})`,
      this.context.operation,
    );
    this.attributesSchema = attributesSchema;
  }

  private collectRelationshipsData(relationshipsSchema: SchemaModel): void {
    for (const { name, schema } of relationshipsSchema.getPropertiesDeep()) {
      let propDataSchema = schema.getProperty('data') ?? null;

      if (!propDataSchema && schema.isComposite()) {
        this.context.logger.warn(`TODO Handle anyOf/oneOf as resource relationship`);
        return;
      }

      let isArray = false;
      if (propDataSchema?.type === 'array') {
        propDataSchema = propDataSchema?.items;
        isArray = true;
      }

      assert(
        propDataSchema,
        `Cannot find items schema for property ${name}`,
        this.context.operation,
      );

      let cardinality = RelationshipCardinality.One;
      if (isArray) {
        cardinality = RelationshipCardinality.Many;
      } else if (propDataSchema.isNullish()) {
        cardinality = RelationshipCardinality.ZeroOrOne;
      }

      let rawResType: JSONValue | undefined;
      if (propDataSchema.isComposite()) {
        const resTypes = getSchemaMultiProperty(propDataSchema, 'type');
        if (resTypes.length !== 1) {
          global.console.log(resTypes);
        }
        assert(
          resTypes.length === 1,
          `Expected only one definition of the 'type' attribute, got ${resTypes.length}. Schema ${
            (relationshipsSchema.parent as SchemaModel).title as string
          }, property ${name}`,
          this.context.operation,
        );
        rawResType = resTypes[0].enum?.at(0);
      } else {
        rawResType = propDataSchema.getPropertyDeep('type')?.enum?.at(0);
      }

      assert(
        rawResType && typeof rawResType === 'string',
        `Resource type must be a string. Got ${String(rawResType)}`,
        this.context.operation,
      );
      const resType = rawResType;

      assert(
        resType && typeof resType === 'string',
        `Precondition failed ${
          this.context.operation.parent.pathUrl
        } prop=${name} ${typeof resType} ${typeof propDataSchema}`,
        this.context.operation,
      );

      this.relationships.set(name, {
        resourceType: resType,
        cardinality,
        required: relationshipsSchema.required.has(name),
      });
    }
  }

  generateCode(generatedTypes: Set<string>): void {
    if (!generatedTypes.has(this.name)) {
      generatedTypes.add(this.name);
    } else {
      return;
    }

    const templateName = this.isRequestBody ? 'JSONAPIClientResource' : 'JSONAPIServerResource';

    addImportDeclaration(this.context.sourceFile, '@fresha/api-tools-core', `t:${templateName}`);

    const typeAlias = addTypeAlias(this.context.sourceFile, this.name, templateName, true);
    const typeAliasType = typeAlias.getNodeProperty('type').asKindOrThrow(SyntaxKind.TypeReference);

    if (this.resourceType) {
      typeAliasType.addTypeArgument(`'${this.resourceType}'`);
    }
    if (this.attributesSchema?.properties.size) {
      const typeLiteral = typeAliasType.addTypeArgument('{}').asKindOrThrow(SyntaxKind.TypeLiteral);
      for (const [name, schema] of this.attributesSchema.properties) {
        const typeName = schemaToType(schema);
        if (!typeName.startsWith('Unknown')) {
          typeLiteral.addProperty({
            kind: StructureKind.PropertySignature,
            name: `'${name}'`,
            type: typeName,
            hasQuestionToken: !this.attributesSchema.required.has(name),
          });
        }
      }
    }
    if (this.relationships.size) {
      if (!this.attributesSchema?.properties.size) {
        typeAliasType.addTypeArgument('{}');
      }

      const typeLiteral = typeAliasType.addTypeArgument('{}').asKindOrThrow(SyntaxKind.TypeLiteral);
      for (const [name, info] of this.relationships) {
        const relName = `JSONAPIResourceRelationship${info.cardinality.toString()}`;
        addImportDeclaration(this.context.sourceFile, '@fresha/api-tools-core', `t:${relName}`);

        typeLiteral.addProperty({
          kind: StructureKind.PropertySignature,
          name: `'${name}'`,
          type: `${relName}<'${info.resourceType}'>`,
          hasQuestionToken: !info.required,
        });
      }
    }
  }
}
