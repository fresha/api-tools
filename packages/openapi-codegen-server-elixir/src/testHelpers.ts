import { Project } from '@fresha/code-morph-ex';
import { createRegistry } from '@fresha/json-api-model';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createTestContext as createBaseTestContext } from '@fresha/openapi-codegen-test-utils';
import { OpenAPIFactory } from '@fresha/openapi-model/build/3.0.3';

import { Generator } from './parts';

import type { Context } from './context';

export const createTestContext = (phoenixApp: string, rootDir = '/'): Context => {
  const openapi = OpenAPIFactory.create();
  const base = createBaseTestContext(openapi, rootDir);

  const project = new Project({
    rootDir,
    phoenixApp,
    useInMemoryFileSystem: true,
  });

  return {
    ...base,
    testObjectFactoryModuleName: `${project.getAppModuleName()}.Factory`,
    openapi,
    project,
    registry: createRegistry(),
  };
};

export const createGenerator = (phoenixApp = 'api_tools_web'): Generator => {
  const context = createTestContext(phoenixApp, '/');
  const generator = new Generator(context);
  return generator;
};
