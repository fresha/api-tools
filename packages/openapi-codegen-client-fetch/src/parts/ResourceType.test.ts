import assert from 'assert';

import { buildEmployeeSchemasForTesting } from '@fresha/openapi-codegen-test-utils';
import {
  addResourceAttributes,
  getOperationDefaultResponseSchemaOrThrow,
  MEDIA_TYPE_JSON_API,
  setResourceSchema,
} from '@fresha/openapi-codegen-utils';
import { OpenAPIFactory, OperationModel } from '@fresha/openapi-model/build/3.0.3';

import { createActionTestContext } from '../testHelpers';

import { NamedType } from './NamedType';
import { ResourceType } from './ResourceType';

import '@fresha/openapi-codegen-test-utils/build/matchers';

const createResourceType = (
  operation: OperationModel,
  name: string,
  isRequestBody = false,
): ResourceType => {
  const context = createActionTestContext(operation, '/src/index.ts');
  const schema = getOperationDefaultResponseSchemaOrThrow(operation, true);
  assert(schema);
  return new ResourceType(context, name, schema, isRequestBody);
};

let operation = OpenAPIFactory.create().setPathItem('/hello').setOperation('get');
let namedTypes = new Map<string, NamedType>();
let generatedTypes = new Set<string>();

beforeEach(() => {
  operation = OpenAPIFactory.create().setPathItem('/hello').setOperation('get');
  namedTypes = new Map<string, NamedType>();
  generatedTypes = new Set<string>();
});

test('simple, client resource', () => {
  operation.setDefaultResponse('test').setContent(MEDIA_TYPE_JSON_API).setSchema(null);

  const resourceType = createResourceType(operation, 'SimpleResource', true);

  namedTypes.set(resourceType.name, resourceType);

  resourceType.collectData(namedTypes);
  resourceType.generateCode(generatedTypes);

  expect(resourceType.context.project.getSourceFileOrThrow('src/types.ts'))
    .toHaveFormattedTypeScriptText(`
    import type { JSONAPIClientResource } from '@fresha/api-tools-core';

    export type SimpleResource = JSONAPIClientResource;
  `);
});

test('attributes', () => {
  const schema = operation
    .setDefaultResponse('test')
    .setContent(MEDIA_TYPE_JSON_API)
    .setSchema(null);
  setResourceSchema(schema, 'hello');
  addResourceAttributes(schema, {
    name: { type: 'string', required: true },
    age: { type: 'integer', nullable: true, required: false },
    gender: { type: 'string', nullable: true, required: true, enum: ['male', 'female', 'other'] },
    active: 'boolean',
  });

  const resourceType = createResourceType(operation, 'HelloResource');

  namedTypes.set(resourceType.name, resourceType);

  resourceType.collectData(namedTypes);
  resourceType.generateCode(generatedTypes);

  expect(resourceType.context.project.getSourceFile('src/types.ts')).toHaveFormattedTypeScriptText(`
    import type { JSONAPIServerResource } from '@fresha/api-tools-core';

    export type HelloResource = JSONAPIServerResource<
      'hello',
      {
        name: string;
        age?: number | null;
        gender: 'male' | 'female' | 'other' | null;
        active?: boolean;
      }
    >;
  `);
});

test('relationships', () => {
  buildEmployeeSchemasForTesting(operation.root);

  const employee = operation.root.components.getSchemaOrThrow('EmployeeResource');

  // remove attributes to unclutter output
  employee.getPropertyDeepOrThrow('attributes').clearProperties();

  const mediaType = operation.setDefaultResponse('test').setContent(MEDIA_TYPE_JSON_API);
  mediaType.schema = employee;

  const resourceType = createResourceType(operation, 'HelloResource');

  namedTypes.set(resourceType.name, resourceType);

  resourceType.collectData(namedTypes);
  resourceType.generateCode(generatedTypes);

  expect(resourceType.context.project.getSourceFile('src/types.ts')).toHaveFormattedTypeScriptText(`
    import type {
      JSONAPIServerResource,
      JSONAPIResourceRelationship1,
      JSONAPIResourceRelationship0,
      JSONAPIResourceRelationshipN,
    } from '@fresha/api-tools-core';

    export type HelloResource = JSONAPIServerResource<
      'employees',
      {},
      {
        manager: JSONAPIResourceRelationship1<'employees'>;
        buddy: JSONAPIResourceRelationship0<'employees'>;
        subordinates: JSONAPIResourceRelationshipN<'employees'>;
      }
    >;
  `);
});
