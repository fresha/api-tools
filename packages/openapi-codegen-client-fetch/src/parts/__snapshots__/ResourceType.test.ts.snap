// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`attributes 1`] = `
"import type { JSONAPIServerResource } from "@fresha/api-tools-core";


/**
 * HelloResource
 *
 * Attributes:
 * -  name
 * -  age
 * -  gender
 * -  active
 *
 */
export type HelloResource = JSONAPIServerResource<'hello', {
        name: string;
        age?: number | null;
        gender: 'male' | 'female' | 'other' | null;
        active?: boolean;
}>;
"
`;

exports[`relationships: src/types.ts 1`] = `
"import type { JSONAPIServerResource, JSONAPIResourceRelationship1, JSONAPIResourceRelationship0, JSONAPIResourceRelationshipN } from "@fresha/api-tools-core";


/**
 * Employee resource
 *
 * This resource contains information about a single employee.
 * It also links to other related resources.
 *
 * Relationships:
 * -  manager relationship to the 'employees' resource
 * -  buddy relationship to the 'employees' resource
 * -  subordinates relationship to the 'employees' resource
 *
 */
export type HelloResource = JSONAPIServerResource<'employees', {}, {
        manager: JSONAPIResourceRelationship1<'employees'>;
        buddy: JSONAPIResourceRelationship0<'employees'>;
        subordinates: JSONAPIResourceRelationshipN<'employees'>;
}>;
"
`;

exports[`simple, client resource: src/types.ts 1`] = `
"import type { JSONAPIClientResource } from "@fresha/api-tools-core";


export type SimpleResource = JSONAPIClientResource;
"
`;
